src/coreclr/tools/Common/Compiler/ObjectWriter/WasmNative.cs

```diff
@@ -0,0 +1,182 @@
+    }
+
+    public static class WasmValueTypeExtensions
+    {
+        public static string ToTypeString(this WasmValueType valueType)
+        {
+            return valueType switch
+            {
+                WasmValueType.I32 => "i32",
+                WasmValueType.I64 => "i64",
+                WasmValueType.F32 => "f32",
+                WasmValueType.F64 => "f64",
+                _ => "unknown",
+            };
+        }
+    }
+
+    #nullable enable
+    public readonly struct WasmResultType : IEquatable<WasmResultType>
+    {
+        private readonly WasmValueType[] _types;
+        public ReadOnlySpan<WasmValueType> Types => _types;
+
+        /// <summary>
+        /// Initializes a new instance of the WasmResultType class with the specified value types.
+        /// </summary>
+        /// <param name="types">An array of WasmValueType elements representing the types included in the result. If null, an empty array is
+        /// used.</param>
+        public WasmResultType(WasmValueType[]? types)
+        {
+            _types = types ?? Array.Empty<WasmValueType>();
+        }
+
+        public bool Equals(WasmResultType other) => Types.SequenceEqual(other.Types);
+        public override bool Equals(object? obj)
+        {
+            return obj is WasmResultType other && Equals(other);
+        }
+
+        public override int GetHashCode()
+        {
+            if (_types == null || _types.Length == 0)
+                return 0;
+
+            int code = _types[0].GetHashCode();
+            for (int i = 1; i < _types.Length; i++)
+            {
+                code = HashCode.Combine(code, _types[i].GetHashCode());
+            }
+
+            return code;
+        }
+
+        public int EncodeSize()
+        {
+            uint sizeLength = DwarfHelper.SizeOfULEB128((ulong)_types.Length);
+            return (int)(sizeLength + (uint)_types.Length);
+        }
+
+        public int Encode(Span<byte> buffer)
+        {
+            int sizeLength = DwarfHelper.WriteULEB128(buffer, (ulong)_types.Length);
+            var rest = buffer.Slice(sizeLength);
+            for (int i = 0; i < _types.Length; i++)
+            {
+                rest[i] = (byte)_types[i];
+            }
+            return (int)(sizeLength + (uint)_types.Length);
+        }
+    }
+
+    public static class WasmResultTypeExtensions
+    {
+        public static string ToTypeListString(this WasmResultType result)
+        {
+            return string.Join(" ", result.Types.ToArray().Select(t => WasmValueTypeExtensions.ToTypeString(t)));
+        }
+    }
+
+    public struct WasmFuncType : IEquatable<WasmFuncType>
+    {
+        private readonly WasmResultType _params;
+        private readonly WasmResultType _returns;
+
+        public WasmFuncType(WasmResultType paramTypes, WasmResultType returnTypes)
+        {
+            _params = paramTypes;
+            _returns = returnTypes;
+        }
+
+        public readonly byte[] Encode()
```

Comment:
Designs that allocate `return byte[]` tend to allocate and copy more than necessary. Why not use the same pattern as for `WasmResultType`, i. e. taking the writer/stream here and writing the bytes?

Reply:
That is a good idea, I have implemented this.